seems a real failure
```
# github.com/cilium/cilium/daemon
daemon/endpoint.go:388:2: unreachable code
make[1]: *** [govet] Error 2
make[1]: Leaving directory `/home/travis/gopath/src/github.com/cilium/cilium'
make: *** [unit-tests] Error 2
```

--- a/daemon/endpoint.go
+++ b/daemon/endpoint.go
@@ -1,4 +1,4 @@
-// Copyright 2016-2019 Authors of Cilium
+// Copyright 2016-2020 Authors of Cilium
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.
@@ -18,6 +18,7 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"net"
 	"sync"
 	"time"
 
@@ -353,12 +354,13 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 	// so we can return from the creation API call.
 	revCh := ep.WaitForPolicyRevision(ctx, 1, nil)
 
+waitForSuccessfulBuild:
 	for {
 		select {
 		case <-revCh:
 			if ctx.Err() == nil {
 				// At least one BPF regeneration has successfully completed.
-				return ep, 0, nil
+				break waitForSuccessfulBuild
 			}
 
 		case <-ctx.Done():
@@ -373,7 +375,7 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 				// return immediately to let the sidecar container start,
 				// in case it is required to enforce L7 rules.
 				logger.Info("Endpoint has sidecar proxy, returning from synchronous creation request before regeneration has succeeded")
-				return ep, 0, nil
+				break waitForSuccessfulBuild
 			}
 		}
 
@@ -381,6 +383,27 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 			return d.errorDuringCreation(ep, fmt.Errorf("timeout while waiting for initial endpoint generation to complete"))
 		}
 	}
+
+	// The endpoint has been successfully created, stop the expiration
+	// timers of all attached IPs
+	if addressing := epTemplate.Addressing; addressing != nil {
+		if uuid := addressing.IPV4ExpirationUUID; uuid != "" {
+			if ip := net.ParseIP(addressing.IPV4); ip != nil {
+				if err := d.ipam.StopExpirationTimer(ip, uuid); err != nil {
+					return d.errorDuringCreation(ep, err)
+				}
+			}
+		}
+		if uuid := addressing.IPV6ExpirationUUID; uuid != "" {
+			if ip := net.ParseIP(addressing.IPV6); ip != nil {
+				if err := d.ipam.StopExpirationTimer(ip, uuid); err != nil {
+					return d.errorDuringCreation(ep, err)
+				}
+			}
+		}
+	}
+
+	return ep, 0, nil
 }
 
 func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder {

--- a/daemon/ipam.go
+++ b/daemon/ipam.go
@@ -1,4 +1,4 @@
-// Copyright 2016-2017 Authors of Cilium
+// Copyright 2016-2020 Authors of Cilium
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.
@@ -18,12 +18,14 @@ import (
 	"fmt"
 	"net"
 	"strings"
+	"time"
 
 	"github.com/cilium/cilium/api/v1/models"
 	ipamapi "github.com/cilium/cilium/api/v1/server/restapi/ipam"
 	"github.com/cilium/cilium/pkg/api"
 	"github.com/cilium/cilium/pkg/cidr"
 	"github.com/cilium/cilium/pkg/datapath"
+	"github.com/cilium/cilium/pkg/defaults"
 	"github.com/cilium/cilium/pkg/endpointmanager"
 	"github.com/cilium/cilium/pkg/ipam"
 	"github.com/cilium/cilium/pkg/logging/logfields"
@@ -47,7 +49,11 @@ func NewPostIPAMHandler(d *Daemon) ipamapi.PostIPAMHandler {
 func (h *postIPAM) Handle(params ipamapi.PostIPAMParams) middleware.Responder {
 	family := strings.ToLower(swag.StringValue(params.Family))
 	owner := swag.StringValue(params.Owner)
-	ipv4Result, ipv6Result, err := h.daemon.ipam.AllocateNext(family, owner)
+	var expirationTimeout time.Duration
+	if swag.BoolValue(params.Expiration) {
+		expirationTimeout = defaults.IPAMExpiration
+	}
+	ipv4Result, ipv6Result, err := h.daemon.ipam.AllocateNextWithExpiration(family, owner, expirationTimeout)
 	if err != nil {
 		return api.Error(ipamapi.PostIPAMFailureCode, err)
 	}
@@ -60,20 +66,22 @@ func (h *postIPAM) Handle(params ipamapi.PostIPAMParams) middleware.Responder {
 	if ipv4Result != nil {
 		resp.Address.IPV4 = ipv4Result.IP.String()
 		resp.IPV4 = &models.IPAMAddressResponse{
-			Cidrs:     ipv4Result.CIDRs,
-			IP:        ipv4Result.IP.String(),
-			MasterMac: ipv4Result.Master,
-			Gateway:   ipv4Result.GatewayIP,
+			Cidrs:          ipv4Result.CIDRs,
+			IP:             ipv4Result.IP.String(),
+			MasterMac:      ipv4Result.Master,
+			Gateway:        ipv4Result.GatewayIP,
+			ExpirationUUID: ipv4Result.ExpirationUUID,
 		}
 	}
 
 	if ipv6Result != nil {
 		resp.Address.IPV6 = ipv6Result.IP.String()
 		resp.IPV6 = &models.IPAMAddressResponse{
-			Cidrs:     ipv6Result.CIDRs,
-			IP:        ipv6Result.IP.String(),
-			MasterMac: ipv6Result.Master,
-			Gateway:   ipv6Result.GatewayIP,
+			Cidrs:          ipv6Result.CIDRs,
+			IP:             ipv6Result.IP.String(),
+			MasterMac:      ipv6Result.Master,
+			Gateway:        ipv6Result.GatewayIP,
+			ExpirationUUID: ipv6Result.ExpirationUUID,
 		}
 	}
 

--- a/pkg/client/ipam.go
+++ b/pkg/client/ipam.go
@@ -26,7 +26,7 @@ const (
 )
 
 // IPAMAllocate allocates an IP address out of address family specific pool.
-func (c *Client) IPAMAllocate(family, owner string) (*models.IPAMResponse, error) {
+func (c *Client) IPAMAllocate(family, owner string, expiration bool) (*models.IPAMResponse, error) {
 	params := ipam.NewPostIPAMParams().WithTimeout(api.ClientTimeout)
 
 	if family != "" {
@@ -36,6 +36,8 @@ func (c *Client) IPAMAllocate(family, owner string) (*models.IPAMResponse, error
 	if owner != "" {
 		params.SetOwner(&owner)
 	}
+
+	params.SetExpiration(&expiration)
 
 	resp, err := c.IPAM.PostIPAM(params)
 	if err != nil {

--- a/pkg/defaults/defaults.go
+++ b/pkg/defaults/defaults.go
@@ -306,4 +306,8 @@ const (
 
 	// RestoreV6Addr is used as match for cilium_host v6 (router) address
 	RestoreV6Addr = "cilium.v6.internal.raw "
+
+	// IPAMExpiration is the timeout after which an IP subject to expiratio
+	// is being released again if no endpoint is being created in time.
+	IPAMExpiration = 3 * time.Minute
 )

--- a/plugins/cilium-cni/cilium-cni.go
+++ b/plugins/cilium-cni/cilium-cni.go
@@ -416,7 +416,7 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 	}
 
 	podName := string(cniArgs.K8S_POD_NAMESPACE) + "/" + string(cniArgs.K8S_POD_NAME)
-	ipam, err = c.IPAMAllocate("", podName)
+	ipam, err = c.IPAMAllocate("", podName, true)
 	if err != nil {
 		err = fmt.Errorf("unable to allocate IP via local cilium agent: %s", err)
 		return
@@ -455,6 +455,7 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 
 	if ipv6IsEnabled(ipam) {
 		ep.Addressing.IPV6 = ipam.Address.IPV6
+		ep.Addressing.IPV6ExpirationUUID = ipam.IPV6.ExpirationUUID
 
 		ipConfig, routes, err = prepareIP(ep.Addressing.IPV6, true, &state, int(conf.RouteMTU))
 		if err != nil {
@@ -467,6 +468,7 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 
 	if ipv4IsEnabled(ipam) {
 		ep.Addressing.IPV4 = ipam.Address.IPV4
+		ep.Addressing.IPV4ExpirationUUID = ipam.IPV4.ExpirationUUID
 
 		ipConfig, routes, err = prepareIP(ep.Addressing.IPV4, false, &state, int(conf.RouteMTU))
 		if err != nil {
@@ -578,20 +580,6 @@ func cmdDel(args *skel.CmdArgs) error {
 		//                         the endpoint is always deleted though, no
 		//                         need to retry
 		log.WithError(err).Warning("Errors encountered while deleting endpoint")
-
-		// Endpoint deletion has failed. We can't be sure whether the
-		// CNI ADD was interrupted in between the IP allocation and the
-		// endpoint creation. In order to guarantee to never leak IP
-		// addresses, delete any IP addresses associated with the pod
-		// being deleted. This task is only done manually on endpoint
-		// delete failure as all IPs of an endpoint are released when
-		// the endpoint is deleted.
-		log.Info("Releasing IPs manually due to inability to delete endpoint")
-		podName := string(cniArgs.K8S_POD_NAMESPACE) + "/" + string(cniArgs.K8S_POD_NAME)
-		err = c.IPAMReleaseIP(podName)
-		if err != nil {
-			log.WithError(err).WithField("pod", podName).Warning("Unable to manually release IPs of pod")
-		}
 	}
 
 	netNs, err := ns.GetNS(args.Netns)

--- a/plugins/cilium-docker/driver/ipam.go
+++ b/plugins/cilium-docker/driver/ipam.go
@@ -116,7 +116,7 @@ func (driver *driver) requestAddress(w http.ResponseWriter, r *http.Request) {
 		family = client.AddressFamilyIPv6
 	}
 
-	ipam, err := driver.client.IPAMAllocate(family, "docker-ipam")
+	ipam, err := driver.client.IPAMAllocate(family, "docker-ipam", false)
 	if err != nil {
 		sendError(w, fmt.Sprintf("Could not allocate IP address: %s", err), http.StatusBadRequest)
 		return

